src/imageselector.cpp

```diff
@@ -12,127 +12,172 @@
+  pathTraverser(pathTraverser)
 {
 }
 
 ImageSelector::~ImageSelector(){}
 
-int ImageSelector::getImageRotation(const std::string& fileName)
+void ImageSelector::setDebugMode(bool debugModeIn) 
+{ 
+  debugMode = debugModeIn;
+}
+
+int ReadExifTag(ExifData* exifData, ExifTag tag, bool shortRead = false)
+{
+  int value = -1;
+  ExifByteOrder byteOrder = exif_data_get_byte_order(exifData);
+  ExifEntry *exifEntry = exif_data_get_entry(exifData, tag);
+
+  if (exifEntry)
+  {
+    if (shortRead)
+    {
+      value = exif_get_short(exifEntry->data, byteOrder);
+    }
+    else
+    {
+      value = exif_get_long(exifEntry->data, byteOrder);
+    }
+  }
+  return value;
+}
+
+ImageDetails ImageSelector::populateImageDetails(const std::string&fileName, const ImageDisplayOptions &baseOptions)
 {
-  int orientation = 0;
+  ImageDetails imageDetails;
+  int orientation = -1;
+  int imageWidth = -1;
+  int imageHeight = -1;
   ExifData *exifData = exif_data_new_from_file(fileName.c_str());
   if (exifData)
   {
-    ExifByteOrder byteOrder = exif_data_get_byte_order(exifData);
-    ExifEntry *exifEntry = exif_data_get_entry(exifData, EXIF_TAG_ORIENTATION);
+    orientation = ReadExifTag(exifData, EXIF_TAG_ORIENTATION, true);
+
+    /*
+    // It looks like you can't trust Exif dimensions, so just forcefully load the file below
+    // try to get the image dimensions from exifData so we don't need to fully load the file
+    imageWidth = ReadExifTag(exifData, EXIF_TAG_IMAGE_WIDTH);
+    if ( imageWidth == -1)
+      imageWidth = ReadExifTag(exifData, EXIF_TAG_PIXEL_X_DIMENSION);
+
+    imageHeight = ReadExifTag(exifData, EXIF_TAG_RELATED_IMAGE_WIDTH); // means height, height is related to width
+    if ( imageHeight == -1)
+      imageHeight = ReadExifTag(exifData, EXIF_TAG_PIXEL_Y_DIMENSION);*/
 
-    if (exifEntry)
-    {
-      orientation = exif_get_short(exifEntry->data, byteOrder);
-    }
     exif_data_free(exifData);
   }
 
   int degrees = 0;
   switch(orientation) {
       case 8:
-      degrees = 270;
-      break;
+        degrees = 270;
+        break;
       case 3:
-      degrees = 180;
-      break;
+        degrees = 180;
+        break;
       case 6:
-      degrees = 90;
-      break;
+        degrees = 90;
+        break;
+      default:
+        break;
+  }
+
+  if (imageWidth <=0 || imageHeight <=0) 
+  {
+    // fallback to QPixmap to determine image size
+    QPixmap p( fileName.c_str() );
+    imageWidth = p.width();
+    imageHeight = p.height();
+  }
+
+  // if the image is rotated then swap height/width here to show displayed sizes
+  if( degrees == 90 || degrees == 270 )
+  {
+    std::swap(imageWidth,imageHeight);
   }
-  return degrees;
+
+  // setup the imageDetails structure
+  imageDetails.filename = fileName;
+
+  imageDetails.width = imageWidth;
+  imageDetails.height = imageHeight;
+  imageDetails.rotation = degrees;
+  if (imageWidth > imageHeight) {
+    imageDetails.aspect = ImageAspect_Landscape;
```

Comment:
can you return here so we don't need that much if-else-if-else logic?

Reply:
Not really, we are update a member here and still changing other parts of the structure. I could make it a helper function but that wouldn't change the if-else-if flow really and wouldn't be reused, so not useful?

Reply:
indeed I think `imagedetails.aspect()` should probably be a function of imageDetails.width and imageDetails.height instead of populating it once here.

Reply:
also I think it's a bit confusing to use `imageAspect_Any` for squares. After reading a bit more I understand you use it as a trick to render the image  in  landscape and portrait mode. However, that means `imageDetails.aspect` doesn't reflect the aspect of the image but _for which aspects to render it_.

Reply:
Probably you could make a method `isValidForScreenAspect()` at the ImageDetais class that does this calculation.